sdk/java/pulumi/src/main/java/com/pulumi/provider/internal/infer/ComponentAnalyzer.java

```diff
@@ -0,0 +1,355 @@
+        }
+        if (type.equals(Boolean.class) || type.equals(boolean.class)) {
+            return BuiltinTypeSpec.BOOLEAN;
+        }
+        return BuiltinTypeSpec.OBJECT;
+    }
+
+    private String getTypeName(Class<?> type) {
+        String name = type.getSimpleName();
+        return name.endsWith("Args") ? name.substring(0, name.length() - 4) : name;
+    }
+
+    private boolean isResourceType(Class<?> type) {
+        return CustomResource.class.isAssignableFrom(type);
+    }
+
+    private static final class TypeAnalysis {
+        private final Map<String, PropertySpec> properties;
+        private final Set<String> required;
+
+        private TypeAnalysis(Map<String, PropertySpec> properties, Set<String> required) {
+            this.properties = properties;
+            this.required = required;
+        }
+
+        public Map<String, PropertySpec> properties() {
+            return properties;
+        }
+
+        public Set<String> required() {
+            return required;
+        }
+    }
+
+    private TypeSpec analyzeTypeParameter(Type type, String context, boolean isOutput) {
+        if (type instanceof ParameterizedType) {
+            ParameterizedType paramType = (ParameterizedType) type;
+            if (paramType.getRawType().equals(Output.class)) {
+                // Analyze inner type, marking it as being in an Output
+                return analyzeTypeParameter(paramType.getActualTypeArguments()[0], context, true);
+            }
+            if (paramType.getRawType().equals(List.class)) {
+                // Propagate isOutput to list items
+                TypeSpec itemSpec = analyzeTypeParameter(paramType.getActualTypeArguments()[0], context, isOutput);
+                return TypeSpec.ofArray(itemSpec);
+            }
+            if (paramType.getRawType().equals(Map.class)) {
+                Type keyType = paramType.getActualTypeArguments()[0];
+                if (!keyType.equals(String.class)) {
+                    throw new IllegalArgumentException(
+                        String.format("map keys must be strings, got '%s' for '%s'", 
+                            ((Class<?>)keyType).getSimpleName(),
+                            context)
+                    );
+                }
+                // Propagate isOutput to map values
+                TypeSpec valueSpec = analyzeTypeParameter(paramType.getActualTypeArguments()[1], context, isOutput);
+                return TypeSpec.ofDict(valueSpec);
+            }
+        }
+        
+        if (type instanceof Class<?>) {
+            Class<?> clazz = (Class<?>) type;
+            if (isBuiltinType(clazz)) {
+                return TypeSpec.ofBuiltin(getBuiltinTypeName(clazz), isOutput ? null : true);
+            } 
+            
+            var specialTypeRef = getSpecialTypeRef(clazz);
+            if (specialTypeRef.isPresent()) {
+                return TypeSpec.ofRef(specialTypeRef.get(), isOutput ? null : true);
```

Comment:
Is this a common pattern for Java to pass `null` for `Boolean`s here rather than false? Iow, couldn't we just pass `!isOutput` here rather than having this ternary operator everywhere?

Reply:
Great question. Honestly, I don’t know what is idiomatic in Java. But my goal here is not to add plain:false into the schema but leave it out instead (for cleanness). I’ll double-check if there is an option to omit default values in serialization.

Reply:
Ah that makes sense.  Thanks for the explanation!